modules/pallets/messaging-fees/src/impls.rs

```diff
@@ -0,0 +1,293 @@
+					.per_byte_fees
+					.iter()
+					.find(|fee| {
+						let hashed_chain_id = keccak_256(&state_machine.encode());
+						fee.state_id == H256(hashed_chain_id)
+					})
+					.map(|fee| fee.per_byte_fee)
+					.unwrap_or(evm_host_param.default_per_byte_fee);
+				fee
+			},
+			SubstrateHostParam(VersionedHostParams::V1(substrate_params)) => {
+				let fee = substrate_params
+					.per_byte_fees
+					.get(state_machine)
+					.cloned()
+					.unwrap_or(substrate_params.default_per_byte_fee);
+				let fee_u128: u128 = fee.into();
+				U256::from(fee_u128)
+			},
+		};
+
+		Some(per_byte_fee)
+	}
+
+	fn process_bridge_rewards(message: &Message, relayer: T::AccountId) -> Result<(), Error<T>> {
+		let mut messages_by_chain: BTreeMap<StateMachine, Vec<IncentivizedMessage>> =
+			BTreeMap::new();
+
+		match message {
+			Message::Request(msg) =>
+				for req in &msg.requests {
+					if IncentivizedRoutes::<T>::get((req.source, req.dest)).is_some() {
+						let request = Request::Post(req.clone());
+						messages_by_chain
+							.entry(req.dest)
+							.or_default()
+							.push(IncentivizedMessage::Request(request));
+					}
+				},
+			Message::Response(msg) => match msg.datagram.clone() {
+				RequestResponse::Request(requests) =>
+					for req in requests {
+						let source_chain = req.source_chain();
+						let dest_chain = req.dest_chain();
+						if IncentivizedRoutes::<T>::get((source_chain, dest_chain)).is_some() {
+							messages_by_chain
+								.entry(dest_chain)
+								.or_default()
+								.push(IncentivizedMessage::Request(req));
+						}
+					},
+				RequestResponse::Response(responses) =>
+					for res in responses {
+						let source_chain = res.source_chain();
+						let dest_chain = res.dest_chain();
+						if IncentivizedRoutes::<T>::get((source_chain, dest_chain)).is_some() {
+							messages_by_chain
+								.entry(dest_chain)
+								.or_default()
+								.push(IncentivizedMessage::Response(res));
+						}
+					},
+			},
+			_ => return Ok(()),
+		};
+
+		for (state_machine, messages) in &messages_by_chain {
+			let bytes_processed: u32 = messages
+				.iter()
+				.map(|msg| match msg {
+					IncentivizedMessage::Request(req) => match req {
+						Request::Post(post) => post.body.len() as u32,
+						Request::Get(_) => 0,
+					},
+					IncentivizedMessage::Response(res) => match res {
+						Response::Post(post) => post.response.len() as u32,
+						Response::Get(get) => get
+							.values
+							.iter()
+							.filter_map(|storage_val| storage_val.value.as_ref())
+							.map(|bytes| bytes.len())
+							.sum::<usize>() as u32,
+					},
+				})
+				.sum();
+
+			TotalBytesProcessed::<T>::mutate(|total| {
+				*total = total.saturating_add(bytes_processed)
+			});
+			let current_total_bytes = TotalBytesProcessed::<T>::get();
+
+			if let Some(per_byte_fee) = Self::get_per_byte_fee(&state_machine) {
+				let dollar_cost = per_byte_fee.saturating_mul(U256::from(bytes_processed));
```

Comment:
Let's do all price calculations in U256 18 decimals,then convert the final result back to u128 12 decimals